### ISSUE TITLE
Speeding up the Windows build time.

### ISSUE THREAD

**Issue by kdschlosser** [openssl/openssl#10752]
I have a suggestion about speeding up the build process.

As you know it does take a while to compile OpenSSL. I am personally using Windows but this same thing holds true across all platforms for the most part. when building using a using a c compiler the process is split into 2 main sections.. compiling the source into objects and then using a linker to bring those objects together. .. This is an over simplification of it, for the purpose of me explaining this it will do.

Now.. when compiling using cl or clang. or which ever c compiler your platform uses this action gets carried out via some form of a make utility or a development program (Visual Studio...).. the biggest issue with make is it only runs a single instance of the compiler at a time. back to back... compilers do not spawn multiple processes to compile so they have no way of utilizing the whole processor. so on a 6 core processor that compiler only has access to 16.666% of the total CPU. and that is all it will ever have access to. This is not a hard thing to overcome. OpenSSL uses Perl to configure the build process. I have not really dove into the exact use of Perl and I am imagining that it is used to create the makefile... I am completely clueless as to what Perl's capabilities are, I am thinking that it is pretty close to Python as far as what it is able to do... If that is the case then why bother with a makefile at all??.. If Perl has a module or library that is equivalent to the subprocess module in Python then you can compile using the whole processor. this is how you would do it.

You would create n+1 threads where n is equal to the number of CPU cores. the list of files that needs to be compiled gets split up between the threads. they each get a groups of file to compile. each thread iterates over the list and calls the compiler. each time the compiler starts it ends up on a different CPU core. The +1 for the thread count is used to also compile it's files.. There are going to be "down times" or periods where that thread is not waiting for the compiler to finish. in order to make use of that time an addition thread needs to be made. Now I know it is not going to use up all of down time.. But it is going to use up most of it. When I use this type of arrangement the CPU use might drop to 99 or 98 for a split second. I would say that is uses 100% of the CPU  for 99.99% of the time it takes to compile. 

I started doing this exact thing if I need to compile an extension module for python.. If using a library from an external source  I spend the time and tear apart the makefile and write a python script that does what the make file does, it just does it  a lot faster then make can..

I am not able to tell you what the exact numbers would be for OpenSSL But from my experience the percentages seem to be really close when I used this type of system on sever different projects.

I am guesstimating the time as I have not measured it.. But say OpenSSL takes 20 minutes to compile... the new time would be pretty close to 1/7th of that time.. so 2.857 minutes.. I find that to be a fairly sizeable decrease..... 

If using Perl makefiles can still be used if you wanted.. just split the makefiles so each one would be responsible for compiling (not linking) their group of files.. have the threads call make. There would be one more makefile that would run the linker and that would get done by the main thread after the others have finished.. If the ability to start processes exists in Perl I would guess that the output from that process can be redirected like in Python. To keep the output nice and clean you pass the output for each call to the main thread, and the main thread would output the data. Since it is sitting there waiting for the other threads to finish may as well make it do something.

If you want an example of what I am trying to explain I can provide one.. it will be written in python  but it will show what I am trying to explain.

I use this approach when compilation times are high can cause an issue.. CI's only provide an hour to "free" accounts... In my use case I need to compile a Python extension x86 and x64 and it has to be compatible with Python versions 2.7, 3.6, 3.7, 3.8.. Now toss in there this needs to be done for Windows, Debian and Darwin, This is going to take a while to complete.

In Python when compiling an extension the compiler and version that was used to compile python should be used when compiling the extension... I know that in Windows there were 3 different msvc versions used when compiling those Python versions.. I do not know if the same holds true with other OS's. It's moot thing anyway because I would need 2 hours just to compile for Windows.

If we can shrink that compile time down to the 2.8 minutes we can finish all 18 in 50 minutes and that is if the same kind of compiler behavior needs to be done on all platforms not just Windows.. which leaves 10 minutes time for whatever else the build program needs to do..

I would go about doing this myself... but.... OpenSSL has a complex build process.. Someone that is more familiar with it would be able to accomplish a task like this in a fraction of the time. Plus It would also be better for all users including the devs of OpenSSL to be able to enjoy the much smaller build times.  I do not know Perl and I know enough about makefiles to be able to port them to python so long as they aren't overly complex. Now OpenSSL's makefile if I remember correctly is a behemoth at 2500+ lines. I would say that's gonna be complex...

I am finishing up a script that will compile c-ares the way I am talking about. For the moment I am only working with getting this running in Windows. BUt for purposes of an example and to provide factual times it will do the trick.

I am helping the developers of pycurl to repair/upgrade their build system which currently does not function properly. writing a cross platform build system is very complex. But then having to do it to external libraries because of a time issue makes it even more difficult, the time that it takes to learn how the build system works for a dependency from someone who is not involved in that deps development is pretty large.

any thoughts on this are appreciated. If someone wants/is willing to provide me with a step by step breakdown of what the makefile is doing I would then be able to write a python script. I do feel that this kind of a thing  should be added to OpenSSL for everyone to be able to enjoy..

Thanks again for taking the time to read the book above.. :smile:
Kevin

**Comment by kdschlosser** [comment]
OK I finished up writing that script for c-ares while it is not as large a gap as I had thought it would be it is still a HUGE improvement.. The gap is not as large as I hoped because the compile times per file are small. I think there are 55 files in total..

I do realize that I have misused the term thread as far as "building using multiple threads" While this is right it is also not right. when a script makes a call to start the compiler it is "spawning" the compiler.. So it is creating a new process. This is the reason why each compiler that is running ends up on a different core.. while multiple threads are being used to make the call it is technically a multi process build

make compiled
12.6539998055 seconds.

compiled using 7 processes
3.71500015259 seconds.

This is a sizeable gain.. If the compile times were greater then 0.230 seconds per file (avg), the time savings would be larger. I can provide the script if someone is interested in seeing how I went about doing this.

**Comment by levitte** [comment]
Modern Unix `make` implementations have the `-j` flag to ask for parallel builds.  For Windows, there's [Jom](https://wiki.qt.io/Jom), which is essentially `nmake` with parallelism added on.

Does that correspond to what you're asking for?

**Comment by beldmit** [comment]
It may be worth looking at ```ccache``` as a solution in case when you have to rebuild the same source code with the same options often.

**Comment by kdschlosser** [comment]
Jom was written to speed up QT builds on Windows. The author write it with the intention of beainf able to read the QT makefile. so it is not 100% in terms of being a replica of nmake. 

It will work with other makefiles other then QT's, however it states simple makefiles have to be used. I am going to hazard a guess that openssl's makefile will not fall into the "simple" category.

I know that caching can also be done with the CI server.. The problem is how would you go about linking the CI's to run one after the other. and also be able to grab the built files.. I do not think there is a way to achieve this.. This is where the issue lies.. I need openssl compiled 6 times just for Windows. it is going to take at least 20 minutes for each build.. with a time limit of only a single hour I would need to be able to link the CI builds together. This is going to get messy if it is even possible.

I am almost done getting the thing ported for Windows.

**Comment by mspncp** [comment]
FWIW: I was able to do a successful parallel build using jom -j 8 on the OpenSSL_1_1_1-stable branch with Visual Studio 2019 on my Notebook, see https://github.com/openssl/openssl/issues/9931#issuecomment-533730276. And I know that @levitte occasionally used jom.

**Comment by levitte** [comment]
> And I know that @levitte occasionally used jom.

Uhmmm, that would be *very* occasionally.  I've tried it twice and had odd failures...  probably because of what @kdschlosser says, that Jom isn't complete enough to be a good replacement.

Either way, I don't think it's reasonable to make an `nmake` replacement within this project.  If someone wants to pick up Jom and make it good enough and release it, that could be a worthy candidate to use.

**Comment by kdschlosser** [comment]
I would never use it even if it does work currently. because a small change made can stop it from working. and since it is not something that is still being developed it has value to me..

Now.. as I say this.. There is another way..

I stumbled upon this repository..
https://github.com/kiyolee/openssl1_1-win-build

the author has spent some time and made VS solutions for VS 2010 to 2019 for the sole purpose of building openssl.

This SHOULD be added to OpenSSL as a submodule. I can verify that the VS2019 does work properly.
I had to give it a little love to get it to run properly Nothing major..
There needs to be some docs for how to get nasm installed into VS. Perl is still needed and runs from inside of VS.. The author did a fantastic job with getting everything to work.

nasm can be installed by using nuget CLI  which can be gotten from 
https://dist.nuget.org/win-x86-commandline/latest/nuget.exe

then navigate to the solution folder that needs to use nasm and run this command
nuget install nasm2 -Version 2.13.3.1

this will install nasm just like it would if you used nuget from inside of VS

Perl is easy because there is a portable version. 


And finally the last piece.. making the thing compile quicker.
There is a way to give VS a kick in the ass so it will compile faster..

https://docs.microsoft.com/en-us/visualstudio/msbuild/using-multiple-processors-to-build-projects?view=vs-2019

TADA!! MSBuild to the rescue. This is a GREAT solution because MSBuild can process solution files from the command line!   HOORAY it can be built speedy like on a CI server... Personally I prefer my system if building for python, only because it is a single build system.. I am going to let the folks over at pycurl know what I have found and they can decide which way they want to go. I still have to poke about libcurl and see what else it needs in order to compile efficiently I believe that OpenSSL was the challenge..

If OpenSSL was linked to the VS solutions in that repo provided some good documentation on how to run the thing. Add in the pieces to make it use multiple processors when compiling. Add in to the solution retrieving Perl and nasm. this is easy stuff to do so far.. once that is done all the user is going to have to do is to click on BUILD!.. there would be a single requirement to build..  Visual Studio.

I am thinking that this could simplify the Windows build even for the developers of OpenSSL. It sure will speed it up.

**Comment by richsalz** [comment]
it shouldn't be a submodule, as it's really a fork of 1.1.d

And the files and directories change a great deal for 3.0

**Comment by kdschlosser** [comment]
I understand that. I am sure if a conversation was started the author might be willing to seperate the VS bits from the OpenSSL components.. Ifg an interest was expressed to him about adding it in some manner to OpenSSL with him having the ability to maintain and update it..  The easiest way to go about doing that would be to add it to OpenSSL as a submodule. this way he would still retain control over the original repo No permissions have to be delegated to him from the OpenSSL repo. and if for some reason It is no longer going to be used it is a simple thing to remove a sub module. Any issues for it will have their own issue tracker that is not connected to OpenSSL...

I am sure there is going to need to be updates made to it.. This is the key part.. He did that all by his lonesome. If he had someone to go over the change in 3.0 I am sure it would be a snap for him to do. He doesn't have reverse engineer the make file or read through the Perl scripts to see what is going on.. That is a HUGE time savings. I consider what he has made there a HUGE accomplishment... There are several projects inside of the solutions. you can compile as an executable.. a shared library, a static library there are even unit tests.

It's a right proper job that must have taken a long time to do. It simply needs a buff shine by adding the things mentioned above to it.. and it's Building on windows with a single click. Now how many projects that have the complexity that OpenSSL has can make that kind of a claim?? I am willing to bet you can count them on your fingers.. That is a HUGE feat. Building on Windows sucks. we all know this. and then trying to build using a CI, for Windows.. that's simply insult to injury...

how about an appveyor script or a travis script where only a single command is needed..

`msbuild openssl.sln /property:Configuration=Release /property:Platform=x64 /maxcpucount:7  /verbosity:normal /nologo`

and that's a wrap!

it will compile in at least half the time. probably closer to a 1/4 of the time. 

I know there is a concern because of 3.0, ask the guy if he would be willing to have his stuff added to OpenSSL.. I would be shocked if he said no.. It would actually be an honor. The only thing that you would have as a requirement would be that the solutions would have to keep pace with the development of OpenSSL. So they would need to support 3.0.. I am sure it took him a longggg time to make those. going from 1.0 to 1.1 was probably not to hard to do.. But from what it sounds like the jump to 3.0 is going to be a total redo.. He may not want to do it again if he has to spend the time to reverse engineer everything. So it would be crucial to have someone that knows the inner working of the build system work with him..

It's not hard stuff to do.. the person that is gong to work with him needs to do it. I am pretty sure all he is going to need to know is what happens to what files. what compiler arguments get passed. what needs to be passed to perl when it is used..

that kind of stuff.,..

**Comment by Mike-Bal** [comment]
Is there a high level documentation of the build process somewhere (i.e. which files have to be compiled with which options on which platform)?. 

If (after the configureation step) it comes down to "compile all files in folders X,Y and Z and add folder W to the include dir" (which would be the optimal case) you can just add a 20 line cmake file that produces any of make, ninja, VS2015,2017,2019... solutions or whatever build system you want that supports parallel compilation. Of course, that wouldn't capture unit tests, but that would already solve the build time "problem" for most consumers with very little effort.

**Comment by DDvO** [comment]
The OpenSSL build time using the native Windows tools `nmake` and `cl` is really terribly slow and inefficient. 

This also affects the project development itself because the CI runs using AppVeyor to build the code and execute tests on Windows are clearly the bottleneck of the CI:
The many Unix-oriented CI runs finish usually within a few minutes while the few Windows CI runs typically are queued for a day(!) until they run, which then takes some half an hour each. 
One can see the current state of the queue including the currently active run at
https://ci.appveyor.com/project/openssl/openssl/history

**Comment by DDvO** [comment]
As mentioned above (well, I did not read all those partly very lengthy posts in detail), parallelizing the compilation such that one object file is built from its source per CPU core would help a lot. 

Unfortunately this is hard to achieve because `nmake` does not support this. Yet `cl` itself has built-in support for compiling multiple source files per call using the available cores. To take advantage of that, the `makefile` (produced by OpenSSL Perl scripts `Configure` etc.) would need to be re-structured, for instance such that all the `*.c` files in each source folder are compiled jointly.

A couple of months back I experimented with a manual adaptation of the `makefile`. This worked and lead to a very significant reduction in build time. I then started implementing/automating this by adapting the Perl scripts and the files in `Configurations/` such as `windows-makefile.tmpl`, but this is non-trivial and got stuck because I had too many other more pressing things to do.

**Comment by MBalszun** [comment]
@DDvO : Just so I understand you correctly. You are suggesting/planning to change the generated make files (for windows in general or just under a separate config switch?) in such a way, that all sources are compiled in a single  cl invocation. That way the compilation can make use of the /MP switch. 
Correct?

As a comment from a random guy on the internet that has to compile openssl from time to time: I think that is a much better approach than implementing an OpenSSL specific build system on top of Perl.

Personally I'd love to see OpenSSL adopt a well known crossplatform build system, but I also understand that this is not an option, because those might not be supported on all platforms that OpenSSL has to be built on (although I beleive the only thing prefenting cmake from e.g. suppoting openVMS is lack of CI infrastructure - the machinery is apparently already in).

I know that maintaining support for multiple parallel build systems is generally problematic. However, I can't but wonder if maintainence wouldn't actually be simpler if you used e.g. CMake (or meson, or whatever) for platforms like Windows, Linux and most forms for cross compilation and maintian the Perl script "just" for the more exotic platforms (although I understand, you will still need pearl to some degree for code generation). 
That way, the vas majority of users would automatically benefit from the features such build systems provide out of the box (e.g. IDE integration, parallel compilation, easy packaging, easy addition of projects specitific / exotic options, conventions that are known by most users etc.) and the Perl based build scripts could probably be simplified by removing support for e.g. msvc and nmake.

Anyway, I only took a cursory look at the scripts an know absolutely nothing about the Openssl development, so this might be completely unreasonable in practice.

Again, the "Fix" of modifying the generated makefiles on windows to make use of /MP would be very welcome.

**Comment by levitte** [comment]
Another idea would be if we could generate a VS project file and then build with `msbuild`...  I haven't researched that possibility at all, so I have no idea how feasible it is.

**Comment by MBalszun** [comment]
>>    (although I beleive the only thing prefenting cmake from e.g. suppoting openVMS is lack of CI infrastructure)

> Incorrect. cmake can currently not be compiled on OpenVMS because the existing C++ compiler is older than C++11, which is the minimum requirement to build cmake.

I was referring to this  issue:

https://gitlab.kitware.com/cmake/cmake/-/issues/3974

but maybe cmake didn't use c++11 back then. 
Anyway. I'm absolutely not trying to push for CMake here - just thinking loudly. I absolutly understand that there a non-trivial problems involved when maintaining two build systems in parallel and as I don't know wnything about the internals of OpenSSL, I completely defer to your judgment of the situation.

>Another "fix" is to use jom, 

Ddin't you write that it didn't work for you? (https://github.com/openssl/openssl/issues/10752#issuecomment-570809108). But sure. If that works - even better. I'll have to try it out the next time this comes up.

**Comment by levitte** [comment]
I did write that jom doesn't work for me at the time, yes. I haven't tried it since, and welcome anyone who'd like to have a closer look and figure out whatever quirks may be needed. Unfortunately ,i have too many other higher prio things to do this myself

**Comment by kdschlosser** [comment]
Python is able to be run on a lot of different platforms. I am not sure of the scope that OpenSSL needs or if Python is able to cover that scope. I have written a Python program to handle compiling on Windows an the major flavors of Linux including Darwin. It handles setting compiler options based on the compiler being used and creates a separate subprocess call, one call for each available core/processor. when all files are compiled then the linker is run.

Visual Studio and the use of MSBuild will allow for compiling "projects" on separate cores/processes. It does not compile different c/cpp files on different cores/processors.

If you have an 8 core machine and you have a Visual Studio solution that has 8 projects, each of the 8 projects has a different number of source files that need to be compiled. If one of those projects has a large number of source files and takes the longest to compile only a single core is going to be used to compile it.

That can be changed but not by using Visual Studio. In order to change that behavior the compiler needs to be called passing an altered list of files to it.


so imagine this is the solution and the 8 projects inside of the solution.
p1: 10 files
p2: 5 files
p3: 6 files
p4: 12 files
p5: 7 files
p6: 6 files
p7: 10 files
p8: 175 files

VS can call msbuild 8 times passing the information needed for each project. ms build is only going to call cl a single time with the list of files to be compiled. cl is only going to run on a single core.

if you eliminate = msbuild from the equation and handle making the calls to cl yourself you could do this.

core1: 29 files, 2 calls to cl, first with the 10 files for p1 then the second call with 19 files from p8
core2: 29 files, 2 calls to cl, first with the 5 files for p1 then the second call with 24 files from p8
core3: 29 files, 2 calls to cl, first with the 6 files for p1 then the second call with 23 files from p8
core4: 29 files, 2 calls to cl, first with the 12 files for p1 then the second call with 17 files from p8
core5: 29 files, 2 calls to cl, first with the 7 files for p1 then the second call with 22 files from p8
core6: 29 files, 2 calls to cl, first with the 6 files for p1 then the second call with 23 files from p8
core7: 29 files, 2 calls to cl, first with the 10 files for p1 then the second call with 19 files from p8
core8: 28 files, 1 call to cl with 28 files

The file lists can also be organized based on compilation time to further optimize the speed. Using this design I have been able to speed up compilation times. On a project that would take over 5 minutes to compile using the traditional way I was able to shrink that time to 45 seconds using the above mechanics. This is on a machine with 6 cores @3.2gHz.

You can also get really crafty and use 3 different CI's to parallel compile everything and then send the compiled files to the fastest of the CI's to be linked. To be able to do this there would need to be a way to pass IP addresses of the CI  instance that handles the linking. This can be done using an e-mail provider like GMail. and that one CI instance that runs the linker would open a socket for connections so the files can be sent. and that same instance sends an e-mail to a GMail account and that e-mail contains the instances IP. The other instances can connect to the GMail acount and read the mail to get the IP so it will know where to open a socket connection to so it can send the files.

For an OpenSource project that is large the cost of a CI might not be possible if there isn't enough incoming funds, Ci's aren't cheap to get the extra time and resources. so ya gotta think outside the box. If you can combine the resources together from multiple Ci's it would make a large difference.


When calling cl.exe there is no need to specifically direct what core it is going to run on. just open a subprocess to a command line (cmd) and call cl.exe. cl will use any core thatt doesn't already have a cl process running on it.  You will want to have a thread for each call made to subprocess so the output and exit can be monitored.


I know my posts are long. I like to explain everything as complete as I can instead of playing the 20 questions game.

**Comment by xclaesse** [comment]
> although I beleive the only thing prefenting cmake from e.g. suppoting openVMS is lack of CI infrastructure - the machinery is apparently already in

Meson could be an option, it only requires python3. Are there platforms supported by OpenSSL but not supported by python3? AFAIK it works on OpenVMS.

**Comment by mspncp** [comment]
> However, regarding @DDvO's idea, that should be possible to do. My first though is to have a look at the function `doobj` in `Configurations/common.tmpl` and have it take an ARRAY ref with all the object files in one go, and have it pass all the source files to be compiled down to `src2obj()` in one go, and let the individual platform decide how to handle that pile. Then, loops like this:
> 
> https://github.com/openssl/openssl/blob/6511f686c2c56d3ed68477981c1a3821bdc2e565/Configurations/common.tmpl#L371-L373
> 
> Can be changed to:
> 
> ```perl
>    doobj([ @objs ], $lib, intent => "lib", attrs => { %attrs }); 
> ```

@levitte, did I miss you saying "I have a branch"? This sounds like a promising idea for a significant build speed-up and is much more straight forward to implement as your alternative plan to construct Visual Studio Projects. (As the proverb says: "A bird in the hand is worth two in the bush.")

**Comment by nhorman** [comment]
There is a current active community effort here https://github.com/openssl/openssl/issues/16812

Which seeks to implement a secondary meson/ninja build system which would be applicable to speeding up windows builds

closing this as a duplicate of that issue

**Comment by jaykrell** [comment]
We see frequent hangs using jom within vcpkg to build OpenSSL.
Triplet to workaround:
```
# Use nmake instead of jom to avoid hang.
if (PORT STREQUAL "openssl")
  set(VCPKG_CONCURRENCY 1)
endif()
```
